packages/editor-ui/src/views/WorkflowHistory.vue

```diff
@@ -1,19 +1,147 @@
 <script setup lang="ts">
+import { saveAs } from 'file-saver';
+import { onBeforeMount, ref, watchEffect } from 'vue';
+import { useRoute, useRouter } from 'vue-router';
+import { VIEWS } from '@/constants';
 import { useI18n } from '@/composables';
+import type { TupleToUnion } from '@/utils/typeHelpers';
+import type { WorkflowHistoryActionTypes, WorkflowHistory } from '@/types/workflowHistory';
+import WorkflowHistoryList from '@/components/WorkflowHistory/WorkflowHistoryList.vue';
+import WorkflowHistoryContent from '@/components/WorkflowHistory/WorkflowHistoryContent.vue';
+import { useWorkflowHistoryStore } from '@/stores/workflowHistory.store';
 
+type WorkflowHistoryActionRecord = {
+	[K in Uppercase<TupleToUnion<WorkflowHistoryActionTypes>>]: Lowercase<K>;
+};
+
+const workflowHistoryActionTypes: WorkflowHistoryActionTypes = [
+	'restore',
+	'clone',
+	'open',
+	'download',
+];
+const WORKFLOW_HISTORY_ACTIONS: WorkflowHistoryActionRecord = workflowHistoryActionTypes.reduce(
+	(record, key) => ({ ...record, [key.toUpperCase()]: key }),
+	{} as WorkflowHistoryActionRecord,
+);
+
+const loadMore = async ({ take }: { take: number }) => {
+	const history = await workflowHistoryStore.getWorkflowHistory(route.params.workflowId, { take });
+	workflowHistoryStore.addWorkflowHistory(history);
+};
+
+const route = useRoute();
+const router = useRouter();
 const i18n = useI18n();
+const workflowHistoryStore = useWorkflowHistoryStore();
+
+const takeItemsAtOnce = ref(20);
+
+onBeforeMount(async () => {
+	await loadMore({ take: takeItemsAtOnce.value });
+
+	if (!route.params.versionId) {
+		await router.replace({
+			name: VIEWS.WORKFLOW_HISTORY,
+			params: {
+				workflowId: route.params.workflowId,
+				versionId: workflowHistoryStore.workflowHistory[0].versionId,
+			},
+		});
+	}
+});
+
+const openInNewTab = (id: WorkflowHistory['versionId']) => {
+	const { href } = router.resolve({
+		name: VIEWS.WORKFLOW_HISTORY,
+		params: {
+			workflowId: route.params.workflowId,
+			versionId: id,
+		},
+	});
+	window.open(href, '_blank');
+};
+
+const downloadVersion = async (id: WorkflowHistory['versionId']) => {
+	const workflowVersion = await workflowHistoryStore.getWorkflowVersion(
+		route.params.workflowId,
+		id,
+	);
+	if (workflowVersion?.workflow) {
+		const { workflow } = workflowVersion;
+		const blob = new Blob([JSON.stringify(workflow, null, 2)], {
+			type: 'application/json;charset=utf-8',
+		});
+		saveAs(blob, workflow.name.replace(/[^a-z0-9]/gi, '_') + '.json');
+	}
+};
+
+const onAction = async ({
+	action,
+	id,
+}: {
+	action: TupleToUnion<WorkflowHistoryActionTypes>;
```

Comment:
`TupleToUnion<WorkflowHistoryActionTypes>` is used heavily throughout the PR and has the disadvantages that it's verbose and opaque as it does not resolve on hover:

<img width="700" alt="Capture 2023-09-27 at 10 05 36@2x" src="https://github.com/n8n-io/n8n/assets/44588767/7e3e48c6-d78b-41ca-9d42-e4c3083e4ea0">

Can we simplify?

```ts
type WorkflowHistoryActionType = WorkflowHistoryActionTypes[number];
```